### PR TITLE
update getSignInUrl() to have consistent return type

### LINKED ISSUE #2693: [DLS] OAuth card SignIn button doesn't work
currently when a bot enables speech channel, the OAuthCard that renders comes with a SignInLink that's retrieved on the bot side. When clicking on SignIn button it only opens up a blank page. I checked the network trace and the SignInLink is returned so WebChat should open the SignIn page.

## Version

> To determine what version of Web Chat you are running, open your browser's development tools, and paste the following line of code into the console.
>
> `[].map.call(document.head.querySelectorAll('meta[name^="botframework-"]'), function (meta) { return meta.outerHTML; }).join('\n')`
>
> If you are using Web Chat outside of a browser, please specify your hosting environment. For example, React Native on iOS, Cordova on Android, SharePoint, PowerApps, etc.

## Steps to reproduce

1. Setup a bot with DLS enabled. Create a dialog in it that uses an OAuthPrompt
2. Talk to the bot using webchat for DLS. make sure the OAuthCard shows up.
3. Click on the SignIn button

## Expected behavior

the browser opens up a new tab which allows user to sign in

## Actual behavior

the browser opens up a new tab with blank page.

## Additional context

[Bug]

### LINKED ISSUE #3439: Inconsistent return types in CardActionContext.getSignInUrl()
Line 118 returns a Promise, but line 125 returns a string or undefined:

https://github.com/microsoft/BotFramework-WebChat/blob/906de7753b140d0925c22e0043628127cc90e070/packages/component/src/Composer.js#L108-L125

This causes an exception in `createDefaultCardActionMiddleware()` where it expects getSignInUrl to return a Promise:

https://github.com/microsoft/BotFramework-WebChat/blob/b47cdac3d10b2c15cde06a3f3be115fe9653724b/packages/component/src/Middleware/CardAction/createCoreMiddleware.js#L75-L79

This issue was discovered while trying to use OAuth + Streaming (specifically DL ASE), though `directLine.getSessionId` may be falsey, the return values should have the same type.

___

### Possible fix:

One solution is to `return Promise.resolve(value);` instead of `return value;`

image

FYI @jiruss, @DDEfromOR, @ckkashyap, @anusharavindrar

### PR DESCRIPTION
**Pull request by stevengum**
> Fixes #3439
> Fixes #2693

## Changelog Entry

-  Fixes [#3439](https://github.com/microsoft/BotFramework-WebChat/issues/3439). Use consistent return type in default CardActionContext.getSignInUrl(), by [@stevengum](https://github.com/stevengum) in PR [#3459](https://github.com/microsoft/BotFramework-WebChat/pull/3459)

## Description

DirectLine instances/objects that do not implement `getSessionId()` results in a codepath that returns a string when the calling code expects a `Promise.resolve()`. This PR wraps the changes the default CardActionContext.getSignInUrl() to always return `Promise.resolve()`, regardless of whether or not the DirectLine object implements `getSessionId()`.

## Specific Changes

- Update **`getSignInUrl()`** to have consistent return type (`Promise<string>`)
- Update warning message when directLine.getSessionId is falsy in the default CardActionContextMiddleware to indicate no-magic-code OAuth flows are not supported
- Add test for CardActionContextMiddleware changes
- Add **`disableNoMagicCode`** option to `__tests__/setup/web/index.html` for testing bugfix
- Add **`getConsoleErrors()`** and **`getConsoleWarnings()`** to `pageObjects`
- Refactor speech.synthesis.js to use **`getConsoleErrors()`** instead of `getConsoleLogs()`
- Add internal helper **`consoleLogFlattener()`** for flattening the data structure returned in `getConsoleErrors()` and `getConsoleWarnings()`
   - `getConsoleLogs()` still returns `Array<Array<string>>` e.g.
     - `[["error", "error-message"], ["warn", "some-warning-message"]]`
   - `getConsoleErrors()` and `getConsoleWarnings()` returns `<Array<string>`, e.g.
     - `["error-message"]` or `["some-warning-message"]`.

## Repro steps
1. Connect to bot using a DirectLineStreaming connection (i.e., DL-ASE, DLS)
1. Begin OAuth flow in bot
1. Click on login button, observe new tab open without a login URL
1. Go back to Web Chat tab, open console and see following error:
image

____

-  ~[ ] I have added tests and executed them locally~ Incomplete
-  [x] I have updated `CHANGELOG.md`
-  ~[ ] I have updated documentation~ Not applicable

## Review Checklist

> This section is for contributors to review your work.

-  [x] Accessibility reviewed (tab order, content readability, alt text, color contrast)
-  [x] Browser and platform compatibilities reviewed
-  [x] CSS styles reviewed (minimal rules, no `z-index`)
-  [x] Documents reviewed (docs, samples, live demo)
-  [x] Internationalization reviewed (strings, unit formatting)
-  [x] `package.json` and `package-lock.json` reviewed
-  [x] Security reviewed (no data URIs, check for nonce leak)
-  [ ] Tests reviewed (coverage, legitimacy)

### CODE REVIEW

**Comment by stevengum** [comment]
@compulim, @corinagum, @cwhitten this PR is ready for review, thanks!

File: __tests__/setup/web/index.html [hunk 1]
```diff
@@ -99,9 +99,7 @@
             payload: {
               activity: {
                 from: { id: 'bot' },
-                id: Math.random()
-                  .toString(36)
-                  .substr(2, 10),
+                id: Math.random().toString(36).substr(2, 10),
```
Comment (stevengum) [inline __tests__/setup/web/index.html:102]: I'm not sure why this change or the one on [L6](https://github.com/microsoft/BotFramework-WebChat/pull/3459/files#diff-d80acb31566c0894d493dc1fa0587ed6R6) slipped by me, but I can back them out if they're unwanted.

I ran `npm run eslint` from the root but it errored out on a shipped Web Chat package untouched by this PR...

Comment (corinagum) [reply __tests__/setup/web/index.html:102]: both are linting fixes, so completely fine to ignore. linter should also run on commit.

**Review by corinagum — APPROVED** [review]
LGTM

**Pull request merged**